### PR TITLE
GAWB-478: optimize get-workspace api by reducing number of queries

### PR DESCRIPTION
**Pull request by davidangb**
Use optimized SQL to calculate workspaceSubmissionStats.

Change get-single-workspace to use the same code path as list-workspaces when calculating workspaceSubmissionStats.

GET /api/workspaces/broad-dsde-dev/large_submissions:
- before: 1655 queries
- after: 23 queries

Important: this only partially addresses GAWB-470. I'd like to push this through and create a follow-on PR to optimize the submissions endpoint.
- [x] **Submitter**: Rebase to develop. DO NOT SQUASH
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Make sure liquibase is updated if appropriate
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [x] Tell  that the PR exists if he wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- [x] **LR**: Initial review by LR and others.
- [x] Comment / review / update cycle:
  - Rest of team may comments on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter updates documentation as needed
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH. **Reassign to LR** for further feedback
- [x]  sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [ ] **Submitter**: Squash commits, rebase if necessary
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Merge to develop 
- [x] **Submitter**: Delete branch after merge
- [x] **Submitter**: Check configuration files in Jenkins in case they need changes
- [x] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [x] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [x] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [x] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 1]
```diff
@@ -322,17 +322,10 @@ class WorkspaceService(protected val userInfo: UserInfo, dataSource: SlickDataSo
...
-      WorkspaceSubmissionStats(
-        lastSuccessDate = workflowsOrderedByDateDesc.find(_.status == WorkflowStatuses.Succeeded).map(_.statusLastChangedDate),
-        lastFailureDate = workflowsOrderedByDateDesc.find(_.status == WorkflowStatuses.Failed).map(_.statusLastChangedDate),
-        runningSubmissionsCount = submissions.count(_.status == SubmissionStatuses.Submitted)
-      )
-    }
+    // listSubmissionSummaryStats works against a sequence of workspaces; we call it just for this one workspace
+    dataAccess.workspaceQuery
+      .listSubmissionSummaryStats(Seq(workspaceContext.workspaceId))
+      .map {p => p.get(workspaceContext.workspaceId).get}
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:328]: I think this change is the one that really makes the difference in this PR.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 2]
```diff
@@ -246,40 +247,63 @@ trait WorkspaceComponent {
...
+      // by using a literal string inside the where clause, we are technically vulnerable to SQL injection.
+      // however, the UUIDs sent in to the literal are generated by the database; the user has no
+      // direct access to this argument
+      val inUUIDs = workspaceIds.map{u => "X'" + u.toString.replace("-","") + "'"}.mkString(",")
+
+      // we abuse UNION here to optimize database interaction. We make one query to return both the
+      // count-of-running-submissions and the last-succeeded/last-failed workflows.
+      // the 'cnt' column will be -1 for workflows, and a value for submissions.
+      import driver.quoteIdentifier
+      val sql:ReadAction[Seq[StatRow]] =
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: what benefit does using raw sql here serve? It seems like it saves 1 query which does not seem worth the complexity.

**Comment by dvoet** [comment]
@davidangb I would prefer not to use raw sql unless it shows significant improvement. I think the bulk of the work is in optimizing the submissions end point so that it does not query per submission or per workflow. I do not believe that raw sql should be required there either. It probably should be somewhere between 1 uber query and 1 query per table that we need info from.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala (hunk 2 above)
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: I'll investigate this once I get back to a dev environment. I thought I saw that Slick generated poor queries for this, but I don't have the info handy and could be remembering wrong.

**Comment by davidangb** [comment]
@dvoet you're right, the plain sql is a micro-optimization here ... I've reverted it. Still, moving to listSubmissionSummaryStats (the surviving change in this PR) is a significant improvement.

**Comment by dvoet** [comment]
agreed :+1:

**Pull request merged**